crates/dbsp/src/dynamic/data.rs

```diff
@@ -0,0 +1,131 @@
+use size_of::SizeOf;
+use std::{fmt::Debug, ops::DerefMut};
+
+use crate::{
+    declare_trait_object, declare_typed_trait_object,
+    dynamic::{
+        storage::SerializeDyn, ArchiveTrait, AsAny, Clonable, ClonableTrait, Comparable,
+        DeserializableDyn, DowncastTrait,
+    },
+    hash::default_hash,
+    DBData, NumEntries,
+};
+
+/// An object-safe trait that defines the minimal set of operations that must be supported by
+/// any value in DBSP.
+///
+/// This trait is object safe and can be invoked via dynamic dispatch.
+pub trait Data:
+    Comparable + Clonable + SerializeDyn + DeserializableDyn + Send + Debug + AsAny + SizeOf
+{
+    /// Compute a hash of the object using default hasher and seed.
+    fn default_hash(&self) -> u64;
+
+    /// Cast any type that implements this trait to `&dyn Data`.
+    ///
+    /// This method will not be needed once trait downcasting has been stabilized.
+    fn as_data(&self) -> &dyn Data;
+
+    /// Cast any type that implements this trait to `&mut dyn Data`.
+    ///
+    /// This method will not be needed once trait downcasting has been stabilized.
+    fn as_data_mut(&mut self) -> &mut dyn Data;
+}
+
+/// The base trait for trait objects that DBSP can compute on.
+pub trait DataTrait: Data + DowncastTrait + ClonableTrait + ArchiveTrait + Eq + Ord {}
+
+impl<Trait: ?Sized> DataTrait for Trait where
+    Trait: Data + DowncastTrait + ClonableTrait + ArchiveTrait + Eq + Ord
+{
+}
+
+// `DynData` is a type alias for `dyn Data`.
+//
+// It is the minimal implementation of trait [`DataTrait`] that only provides the
+// methods required by this trait and nothing else.
+declare_trait_object!(DynData<> = dyn Data<>
+where
+);
+
+// Fixme: we always treat `DynData` as a scalar.
+impl NumEntries for DynData {
+    const CONST_NUM_ENTRIES: Option<usize> = Some(1);
+
+    #[inline]
+    fn num_entries_shallow(&self) -> usize {
```

Comment:
I find it useful for profiling.  It should be implemented correctly where it actually matters. Ideally it would be a bound on `DBData`, but that may make it even harder to use external library types in DBSP.